libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+
+/// This tracer does not currently support thread-local tracers or concurrency controls
+/// Functions should only be used in a single-threaded context
+///
+/// Best Practices:
+///   If possible always avoid creating raw ProgressSpans from ProgressTracer
+///   If possible always use ProgressScopes to handle ProgressSpans
+///   Only use 1 ProgressTracer in an execution, so they should be created at entry points
+///   Use ProgressScope's RAII to handle early returns (i.e. due to errors)
+///   Raw ProgressSpans should be used for special scenarios like tracing asynchronous calls
+///
+/// Notes:
+///   SharedMemSys and DistMemSys will initialize the global ProgressTracer to a JsonTracer
+///     later it will be set to a No-op Tracer, on Fini or the destructor they will call
+///     Close on the global ProgressTracer
+///   This prevents GetProgressTracer() from returning nullptr and ensures Tracers are closed
+///
+///   ProgressScope's will only close their ProgressSpan when their ProgressSpan is the active span
+///   This means that if a user exclusively uses ProgressScopes,
+///   Then parent ProgressSpans will not be closed until their child ProgressSpan's close
+///   i.e.
+///   {
+///      auto scope1 = tracer.StartActiveSpan("1");
+///      auto scope2 = tracer.StartActiveSpan("2");
+///      if (err) { return; }
+///      scope2.Close();
+///      auto scope3 = tracer.StartActiveSpan("3");
+///   }
+///   Always results in scope2's ProgressSpan finishing before scope1's
+///   and scope3's ProgressSpan finishing before scope1's if there is no error
+
+namespace katana {
+
+// This is needed to properly handle being passed const char* values
+// Once we submodule OpenTracing this will be removed and their Value class used instead
+// TODO(Patrick)
+using variant_type = std::variant<
+    bool, int, int64_t, uint32_t, uint64_t, float, double, std::string,
+    const char*>;
+class Value : public variant_type {
+public:
+  Value(bool x) noexcept : variant_type(x) {}
+  Value(int x) noexcept : variant_type(static_cast<int64_t>(x)) {}
+  Value(int64_t x) noexcept : variant_type(x) {}
+  Value(uint32_t x) noexcept : variant_type(static_cast<uint64_t>(x)) {}
+  Value(uint64_t x) noexcept : variant_type(x) {}
+  Value(double x) noexcept : variant_type(x) {}
+  Value(std::string x) noexcept : variant_type(x) {}
+  Value(const char* s) noexcept : variant_type(std::string(s)) {}
+};
+using Tags = std::vector<std::pair<std::string, Value>>;
+
+class ProgressScope;
+class ProgressSpan;
+class ProgressContext;
+
+class KATANA_EXPORT ProgressTracer {
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
+
+  ProgressScope SetActiveSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
```

Comment:
Since this is a virtual base class, probably delete move and copy special member functions.